### DISCUSSION TITLE
Close any pending subscriptions when changing client in context

### DISCUSSION THREAD

**Discussion by jld-adriano** [trpc/trpc#1857] [Q&A] [answered]
For our application, we have a use case where the URL for the websocket api can change, and so our solution has been to treat the trpc client as state, and change it whenever the url changes. However, whenever we change it, there are pending connections that stay in the background. 

Is this use case supported?
If not, is there some easy workaround we could do by calling internals?

Thanks

**Comment by KATT** [comment]
I think there's a close fn on the wsClient

Dig around here https://github.com/trpc/trpc/blob/main/packages/client/src/links/wsLink.ts

That file is probably the one I'm the least proud of in the whole codebase

**Reply by jld-adriano** [reply]
Dug around in the code trying to fix it for our use case. just calling close doesn't seem to be enough, I believe something is "restarting" the connection after close

**Comment by jlalmes** [comment]
Not sure I fully understand your use-case for changing the websocket URL on the fly, but here's an idea:

```ts
// useTrpcWsClient.ts
import { createTRPCClient } from '@trpc/client';
import { createWSClient, wsLink } from '@trpc/client/links/wsLink';
import { useEffect, useMemo } from 'react';

const useTrpcWsClient = (url: string) => {
  const { client, wsClient } = useMemo(() => {
    const wsClient = createWSClient({ url });
    const client = createTRPCClient({
      links: [wsLink({ client: wsClient })],
    });
    return { client, wsClient };
  }, [url]);

  useEffect(() => {
    return () => {
      wsClient.close();
    }
  }, [wsClient]);

  return client;
};

export default useTrpcWsClient;
```

**Reply by jld-adriano** [reply]
Yeah this is our basic implementation (but with a ref to close the last one, instead of this one), but the close doesn't work as expected, the connections keep requesting in the background

**Reply by jlalmes** [reply]
Could you please provide a reproduction.

**Reply by jld-adriano** [reply]
Yeah I'll look into making a minimal reproduction

**Reply by jld-adriano** [reply]
https://github.com/jld-adriano/trpc-discussion-1857-reproduction Here's the minimal repro, just start both app and server, change the host in the input, and you'll see both websocket connections keep trying to connect

**Reply by jld-adriano** [reply]
@jlalmes Thought you might have missed this comment :) we are still using the workaround, it's working fine for our use case

**Reply by jlalmes** [reply]
Hi @jld-adriano, sorry for the delay - I've just checked out the reproduction (thanks!) and can confirm this is a bug. Will ship a fix shortly.

**Reply by jld-adriano** [reply]
Thanks a lot! Can you tag me when you do it so I can give feedback if it fixes our issue. Cheers

**Comment by jld-adriano** [comment]
In our use case, the socket connection "fails" until we find the right url, and then it works. 
It's for network discovery purposes.

**Reply by jld-adriano** [reply]
As a workaround, we are returning  999999999 as a retry interval when the client changes, for all the old clients

**Comment by KATT** [comment, marked as answer]
@jld-adriano this should be fixed in `9.24.0`, let us know! 🙏

**Reply by jlalmes** [reply]
Just confirmed reproduction is working as intended once upgrading to `v9.24.0`.
